docs/concepts/gas.md

```diff
@@ -4,10 +4,159 @@ title: Gas
+
+    action_receipt_creation_config: {
+      send_sir:     108059500000,
+      send_not_sir: 108059500000,
+      execution:    108059500000
+    }
+
+The appropriate `send` amount for creating this receipt is also immediately deducted from your account.
+
+The "create account" action won't be finalized until the next block. At this point, the `execution` amount for each of these actions will be deducted from your account (something subtle: the gas units on this next block could be multiplied by a gas price that's up to 1% different, since gas price is recalculated on each block). Adding it all up to find the total transaction fee:
+
+    (create_account_cost.send_sir  + action_receipt_creation_config.send_sir ) * gas_price_at_block_1 +
+    (create_account_cost.execution + action_receipt_creation_config.execution) * gas_price_at_block_2
+</blockquote>
+
+## Costs of complex actions
+
+The numbers above should give you the sense that transactions on NEAR are cheap! But they don't give you much sense of how much it will cost to use a more complex app or operate a NEAR-based business. Let's cover some more complex gas calculations: deploying contracts and function calls.
+
+### Deploying Contracts
+
+The [basic action costs](https://github.com/nearprotocol/nearcore/blob/49b4fcdc297a609d8bb38858fdbf71e7d821f1f5/neard/res/genesis_config.json#L57-L120) include two different values for deploying contracts. Simplified, these are:
+
+    deploy_contract_cost: 184765750000,
+    deploy_contract_cost_per_byte: 6812999,
+
+The first is a baseline cost, no matter the contract size. Keeping in mind these each need to be multiplied by two, for both `send` and `execute` costs, and will also require sending & executing a receipt (see blue box above), the gas units comes to:
+
+    2 * 184765750000 +
+    2 * contract_size_in_bytes * 6812999 +
+    2 * 108059500000
+
+(Divide the resulting number by 10<super>12</super> to get to TGas!)
+
+Note that this covers the cost of uploading and writing bytes to storage, but does _not_ cover the cost of holding these bytes in storage. Long-term storage is compensated via [storage staking], a recoverable cost-per-byte amount that will also be deducted from your account during contract deployment.
+
+The AssemblyScript contract in [this example Fungible Token](https://github.com/near-examples/FT/pull/42) compiles to just over 16kb (the Rust contract is much larger, but this [will be optimized](???)). Using the calculation above, we find that it requires **0.81 TGas** (and thus 0.081mN at minimum gas price) for the transaction fee to deploy the contract, while 1.5N will be locked up for storage staking.
```

Comment:
@nearmax is there an existing issue for tracking the Rust contract size optimization?

Reply:
https://github.com/near/near-sdk-rs/issues/167